### PR TITLE
Shader debug instrumentation

### PR DESCRIPTION
**Pull request by jeffnn**
This is a big one. See the new cpp file for a large comment describing how it works. That's a good place to start before code reviewing.

### CODE REVIEW

**Comment by jeffnn** [comment]
Oh. I forgot to convert back to {-at-end-of-line style.
Hold on...

**Comment by jeffnn** [comment]
Uh, I also forgot to write any tests!
That'll take a while... I'll send mail when I'm ready.

File: lib/HLSL/DxilDebugInstrumentation.cpp [hunk 1]
```diff
@@ -0,0 +1,723 @@
+///////////////////////////////////////////////////////////////////////////////
+//                                                                           //
+// DxilOutputColorBecomesConstant.cpp                                        //
```
Comment (marcelolr) [inline lib/HLSL/DxilDebugInstrumentation.cpp]: File name in banner ...

File: lib/HLSL/DxilDebugInstrumentation.cpp [hunk 2]
```diff
@@ -0,0 +1,723 @@
...
+    IRBuilder<> & Builder;
+  };
+
+  uint32_t m_RemainingReservedSpaceInBytes = 0;
+  Value * m_CurrentIndex = nullptr;
+
+public:
+  static char ID; // Pass identification, replacement for typeid
+  explicit DxilDebugInstrumentation() : ModulePass(ID) {}
+  const char *getPassName() const override { return "DXIL Force Early Z"; }
```
Comment (marcelolr) [inline lib/HLSL/DxilDebugInstrumentation.cpp]: Update pass name.

File: lib/HLSL/DxilDebugInstrumentation.cpp [hunk 3]
```diff
@@ -0,0 +1,723 @@
...
+  void applyOptions(PassOptions O) override;
+  bool runOnModule(Module &M) override;
+
+private:
+  SystemValueIndices addRequiredSystemValues(BuilderContext & BC);
```
Comment (marcelolr) [inline lib/HLSL/DxilDebugInstrumentation.cpp]: If you need to touch this up anyway - code convention in this part of the codebase is "BuilderContext &BC", with no space between ampersand and 'BC'

File: lib/HLSL/DxilDebugInstrumentation.cpp [hunk 4]
```diff
@@ -0,0 +1,723 @@
...
+
+}
+
+bool DxilDebugInstrumentation::runOnModule(Module &M) {
+  DxilModule &DM = M.GetOrCreateDxilModule();
+  LLVMContext & Ctx = M.getContext();
+  OP *HlslOP = DM.GetOP();
+
+  // First record pointers to all instructions in the function:
+  std::vector<Instruction*> AllInstrucitons;
```
Comment (marcelolr) [inline lib/HLSL/DxilDebugInstrumentation.cpp]: Typo: AllInstrucitons -> AllInstructions

File: lib/HLSL/DxilDebugInstrumentation.cpp [hunk 5]
```diff
@@ -0,0 +1,723 @@
...
+  auto & Blocks = DM.GetEntryFunction()->getBasicBlockList();
```
Comment (marcelolr) [inline lib/HLSL/DxilDebugInstrumentation.cpp]: inst_iterator is your friend: http://llvm.org/docs/ProgrammersManual.html#iterating-over-the-instruction-in-a-function

File: lib/HLSL/DxilDebugInstrumentation.cpp [hunk 6]
```diff
@@ -0,0 +1,723 @@
...
+    }
+    else {
+      addStepEntryForType<uint32_t>(DebugShaderModifierRecordTypeDXILStepUint32, BC, Inst);
+    }
+    break;
+  case Type::TypeID::DoubleTyID:
+    addStepEntryForType<double>(DebugShaderModifierRecordTypeDXILStepDouble, BC, Inst);
+    break;
+  case Type::TypeID::FP128TyID:
+  case Type::TypeID::HalfTyID:
```
Comment (marcelolr) [inline lib/HLSL/DxilDebugInstrumentation.cpp]: HalfTyID will show up with true half support (and you may already see this with min16float as well).

File: lib/HLSL/DxilDebugInstrumentation.cpp [hunk 7]
```diff
@@ -0,0 +1,723 @@
...
+    });
+
+    m_RemainingReservedSpaceInBytes -= 4;
+    assert(m_RemainingReservedSpaceInBytes < 1024);  // check for underflow
+
+    if (m_RemainingReservedSpaceInBytes != 0) {
+      m_CurrentIndex = BC.Builder.CreateAdd(m_CurrentIndex, BC.HlslOP->GetU32Const(4));
+    }
+    else {
+        m_CurrentIndex = nullptr;
```
Comment (marcelolr) [inline lib/HLSL/DxilDebugInstrumentation.cpp]: odd indent here

File: lib/HLSL/DxilDebugInstrumentation.cpp [hunk 8]
```diff
@@ -0,0 +1,723 @@
...
+    ParameterTestResult = addPixelShaderProlog(BC, SVIndices);
+    break;
+    case DXIL::ShaderKind::Vertex:
+      ParameterTestResult = addVertexShaderProlog(BC, SVIndices);
+    break;
+    case DXIL::ShaderKind::Compute:
+      ParameterTestResult = addComputeShaderProlog(BC);
+    break;
+  default:
+    assert(false);
```
Comment (marcelolr) [inline lib/HLSL/DxilDebugInstrumentation.cpp]: This is a tiny bit odd. Should we really simply assert if the wrong shader model was picked? Maybe throw hlsl exception with an invalid argument or a new error code?

File: tools/clang/test/HLSL/pix/DebugVSParameters.hlsl [hunk 9]
```diff
@@ -0,0 +1,17 @@
+// RUN: %dxc -Emain -Tvs_6_0 %s | %opt -S -hlsl-dxil-debug-instrumentation,parameter0=1,parameter1=2 | %FileCheck %s
+
+// Check that the instance and vertex id are parsed and present:
+
+// CHECK: %PIX_DebugUAV_Handle = call %dx.types.Handle @dx.op.createHandle(i32 57, i8 1, i32 0, i32 0, i1 false)
+// CHECK: %VertId = call i32 @dx.op.loadInput.i32(i32 4, i32 0, i32 0, i8 0, i32 undef)
+// CHECK: %InstanceId = call i32 @dx.op.loadInput.i32(i32 4, i32 1, i32 0, i8 0, i32 undef)
+// CHECK: %CompareToVertId = icmp eq i32 %VertId, 1
```
Comment (marcelolr) [inline tools/clang/test/HLSL/pix/DebugVSParameters.hlsl]: Are you getting any trouble in release builds with names here?

File: tools/clang/test/HLSL/pix/DebugBasic.hlsl [hunk 10]
```diff
@@ -0,0 +1,29 @@
+// RUN: %dxc -Emain -Tps_6_0 %s | %opt -S -hlsl-dxil-debug-instrumentation | %FileCheck %s
+
+// Check that the basic starting header is present:
```
Comment (marcelolr) [inline tools/clang/test/HLSL/pix/DebugBasic.hlsl]: There should be a test for the case where SV_Position is already in the shader and doesn't have to be added, but simply reused.

File: tools/clang/test/HLSL/pix/DebugVSParameters.hlsl (hunk 9 above)
Comment (jeffnn) [reply tools/clang/test/HLSL/pix/DebugVSParameters.hlsl]: Release builds worked fine. These are names I asserted myself, so they don't get re-used and appended-with-numbers. Not that I got release correct first time: I had to modify some of the checks to remove auto-generated numerical names for variables and labels.

File: tools/clang/test/HLSL/pix/DebugBasic.hlsl (hunk 10 above)
Comment (jeffnn) [reply tools/clang/test/HLSL/pix/DebugBasic.hlsl]: Good idea. I have that on the PIX side, but it'd be good here too.

File: lib/HLSL/DxilDebugInstrumentation.cpp (hunk 1 above)
Comment (jeffnn) [reply lib/HLSL/DxilDebugInstrumentation.cpp]: fixed...

File: lib/HLSL/DxilDebugInstrumentation.cpp (hunk 2 above)
Comment (jeffnn) [reply lib/HLSL/DxilDebugInstrumentation.cpp]: fixed...

File: lib/HLSL/DxilDebugInstrumentation.cpp (hunk 3 above)
Comment (jeffnn) [reply lib/HLSL/DxilDebugInstrumentation.cpp]: fixed... (and a couple of other things like it)

File: lib/HLSL/DxilDebugInstrumentation.cpp (hunk 7 above)
Comment (jeffnn) [reply lib/HLSL/DxilDebugInstrumentation.cpp]: fixed...

File: lib/HLSL/DxilDebugInstrumentation.cpp (hunk 4 above)
Comment (jeffnn) [reply lib/HLSL/DxilDebugInstrumentation.cpp]: fixed...

File: lib/HLSL/DxilDebugInstrumentation.cpp (hunk 5 above)
Comment (jeffnn) [reply lib/HLSL/DxilDebugInstrumentation.cpp]: Thanks... changed

File: lib/HLSL/DxilDebugInstrumentation.cpp (hunk 6 above)
Comment (jeffnn) [reply lib/HLSL/DxilDebugInstrumentation.cpp]: Thanks. I had intended to fill that one out and just plain forgot! Fixed.

File: lib/HLSL/DxilDebugInstrumentation.cpp (hunk 8 above)
Comment (jeffnn) [reply lib/HLSL/DxilDebugInstrumentation.cpp]: Fixed as discussed

File: tools/clang/test/HLSL/pix/DebugBasic.hlsl (hunk 10 above)
Comment (jeffnn) [reply tools/clang/test/HLSL/pix/DebugBasic.hlsl]: Done.

**Pull request merged**